File: docs/execution/test-parallel.php
<?php

$payload = file_get_contents("forxy-payload.json");
$json_obj = json_decode($payload, true);

$multi_handler = curl_multi_init();
$curl_handlers = [];

foreach ($json_obj['requests'] as $id => $req) {
    $ch = curl_init($req['url']);
    
    $json_body = json_encode($req['body']);
    
    curl_setopt($ch, CURLOPT_HTTPGET, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $json_body);
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        'Content-Type: application/json'
    ]);

    curl_multi_add_handle($multi_handler, $ch);
    $curl_handlers[$id] = $ch;
}

$running = null;
$start = microtime(true);
do {
    curl_multi_exec($multi_handler, $running);
} while ($running > 0);

foreach ($curl_handlers as $id => $ch) {
    $response = curl_multi_getcontent($ch);
    $http_status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    echo ("Request $id: Status code: $http_status \n");
    
    curl_multi_remove_handle($multi_handler, $ch);
    curl_close($ch);
}

$end = microtime(true);
curl_multi_close($multi_handler);

echo ("Execution time: " . ($end-$start));

File: docs/execution/test-parallel.py
import requests
import json
import concurrent.futures
import time

with open("forxy-payload.json", 'r') as file:
    payload = json.loads(file.read())

def send_request(req_id, req_data):
    try:
        response = requests.get(req_data["url"], json=req_data["body"])
        print(f"Request {req_id}: Status code: {response.status_code}, Response: {response.text}")
    except Exception as e:
        print(f"Request {req_id} failed: {e}")

start = time.time()

with concurrent.futures.ThreadPoolExecutor() as executor:
    futures = [executor.submit(send_request, req_id, req_data) for req_id, req_data in payload['requests'].items()]

    for future in concurrent.futures.as_completed(futures):
        pass

end = time.time()
print("Execution time: " + str(end-start))
